wordpress-mobile/WordPress-Android
# Add Author button to post settings

Fixes #4230 

- This adds an author button to the post settings screen and the author of the post can be changed with it.
- For sites with many authors (>500 authors), fetching authors can take some time (3-10 seconds). "Loading..." message will appear in the dialog while fetching authors.
- https://developer.wordpress.com/docs/api/1.1/get/sites/%24site/users/ endpoint fails when the `number` parameter is big (500 or greater). I set the `number` parameter to 100. For example, for a site with 500 authors, the app makes 5 calls.

**Possible enhancement for a future task:** Fetched authors can be cached. If we cache authors, we can hide the author button for singlet author sites, and we don't need a loading screen while fetching authors.

***Merge first:***
- [x] Merge https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2523 first.

<img src="https://user-images.githubusercontent.com/2471769/188285273-63fb323a-094a-4e53-bda9-ace8f1d22a09.png" height=540> <img src="https://user-images.githubusercontent.com/2471769/189162399-a6be4676-bbf7-416d-b8ca-ada5d57b04ac.png" height=540> <img src="https://user-images.githubusercontent.com/2471769/188285274-daaed61f-9982-4170-9ace-5af76a0c1b25.png" height=540>

**To test single author site:**
1. Select a site with single author.
2. Tap the blue button at the bottom right of the screen, then "Blog Post" or "Site page" to open the new post screen.
3. Tap the options button at the top right of the screen, then "Post settings".
4. Ensure the post is published with the selected author.

**To test multiple authors site:**
1. Select a site with few authors (2 or 3).
2. Tap the blue button at the bottom right of the screen, then "Blog Post" or "Site page" to open the new post screen.
3. Tap the options button at the top right of the screen, then "Post settings".
4. Ensure the is "Author" button on the Post Settings screen.
5. Tap the "Author" button. Ensure a dialog with author list appears.
6. Change the author. Publish the change.
7. Ensure the post is published with the selected author.
8. Repeat 1-7 with a site with many authors (>500 authors). 

**To test edit post:**
1. Launch the app.
2. Tap My Site → Posts.
3. Tap "Edit" button on an old post. (If there isn't any post, create one first)
3. Tap the options button at the top right of the screen, then "Post settings".
4. Ensure the is "Author" button on the Post Settings screen.
5. Tap the "Author" button. Ensure a dialog with author list appears.
6. Change the author. Publish the change.

## Regression Notes
1. Potential unintended areas of impact
People screen on My Site → MENU tab.

10. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested manually.

11. What automated tests I added (or what prevented me from doing so)
Updated `PublishSettingsViewModelTest`.

PR submission checklist:
